lib/luacrypto_aead.c

```diff
@@ -0,0 +1,260 @@
+	lua_pushinteger(L, crypto_aead_authsize(tfm));
+	return 1;
+}
+
+static inline struct aead_request *luacrypto_aead_newrequest(lua_State *L, const char **combined, u8 **iv,
+	size_t *combined_len, size_t *aad_len, size_t *crypt_len, unsigned int *authsize)
+{
+	struct crypto_aead *tfm = luacrypto_aead_check(L, 1);
+
+	size_t iv_len;
+	const char *l_iv = luaL_checklstring(L, 2, &iv_len);
+	luaL_argcheck(L, iv_len == crypto_aead_ivsize(tfm), 2, "incorrect IV length");
+
+	*combined = luaL_checklstring(L, 3, combined_len);
+
+	*aad_len = (size_t)luaL_checkinteger(L, 4);
+	luaL_argcheck(L, *aad_len >= 0 && *aad_len <= *combined_len, 4, "AAD length out of bounds");
+
+	*crypt_len = *combined_len - *aad_len;
+
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+	struct aead_request *request = lunatik_checknull(L, aead_request_alloc(tfm, gfp));
+	*iv = (u8 *)lunatik_checkalloc(L, iv_len);
+	memcpy(*iv, l_iv, iv_len);
+	*authsize = crypto_aead_authsize(tfm);
+
+	return request;
+}
+
+static inline void luacrypto_aead_setrequest(lua_State *L, struct aead_request *request, const char *combined, u8 *iv,
+	size_t combined_len, size_t aad_len, size_t crypt_len, char *buffer, size_t buffer_len)
+{
+	memcpy(buffer, combined, combined_len);
+
+	struct scatterlist sg_work;
+	sg_init_one(&sg_work, buffer, buffer_len);
+
+	aead_request_set_ad(request, aad_len);
+	aead_request_set_crypt(request, &sg_work, &sg_work, crypt_len, iv);
+	aead_request_set_callback(request, 0, NULL, NULL);
+}
+
+static inline void luacrypto_aead_freerequest(struct aead_request *request, u8 *iv)
+{
+	aead_request_free(request);
+	lunatik_free(iv);
+}
+
+#define LUACRYPTO_AEAD_CHECK_ENCRYPT(L, ix, crypt_len, authsize)
+#define LUACRYPTO_AEAD_CHECK_DECRYPT(L, ix, crypt_len, authsize)	\
+	luaL_argcheck(L, crypt_len >= authsize, ix, "input data (ciphertext+tag) too short for tag")
+
+#define LUACRYPTO_AEAD_LEN_ENCRYPT(combined_len, authsize)	(combined_len + authsize)
+#define LUACRYPTO_AEAD_LEN_DECRYPT(combined_len, authsize)	(combined_len)
+
+#define LUACRYPTO_AEAD_CRYPT_FN(name, NAME, res_factor)									\
+static int luacrypto_aead_##name(lua_State *L) {									\
+	const char *combined;												\
+	u8 *iv;														\
+	unsigned int authsize;												\
+	size_t combined_len, aad_len, crypt_len;									\
+	struct aead_request *request = luacrypto_aead_newrequest(L, &combined, &iv, &combined_len,			\
+		&aad_len, &crypt_len, &authsize); 									\
+															\
+	LUACRYPTO_AEAD_CHECK_##NAME(L, 3, crypt_len, authsize);								\
+	size_t buffer_len = LUACRYPTO_AEAD_LEN_##NAME(combined_len, authsize);						\
+															\
+	luaL_Buffer B;													\
+	char *buffer = luaL_buffinitsize(L, &B, buffer_len);								\
+	luacrypto_aead_setrequest(L, request, combined, iv, combined_len, aad_len, crypt_len, buffer, buffer_len);	\
+	int ret = crypto_aead_##name(request);										\
+	luacrypto_aead_freerequest(request, iv);									\
+	if (ret < 0)													\
```

Comment:
Wouldn't it be the case for lunatik_tryret?

Reply:
I don’t think so: this would lead to that code:
```C
lunatik_tryret(L, ret, crypto_aead_##name, request);
luacrypto_aead_freerequest(request, iv);
```
Which would leak `request` and `iv` if `crypto_aead_##name` fails.